### PR TITLE
Draft documentation for final attributes

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Draft work towards #5598

@JukkaL you can either push directly to this branch or merge this and polish it afterwards.

### CODE REVIEW

**Comment by gvanrossum** [comment]
Do you want this cherry-picked in the 0.640 release branch?

**Comment by JukkaL** [comment]
I think that we can cherry-pick this if we have a non-draft version before the release. I'll finish this up tomorrow.

File: docs/source/final_attrs.rst [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+   import uuid
+   from typing_extensions import Final
+
+   class Snowflake:
+       """An absolutely unique object in the database"""
+       def __init__(self) -> None:
+           self.id: Final = uuid.uuid4()
+
+   # 1000 lines later
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: This comment feels irrelevant.

File: docs/source/final_attrs.rst [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+   class User(Snowflake):
+       id = uuid.uuid4()  # Error: can't override a final attribute
+
+Some other use cases might be solved by using ``@property``, but note that
+neither of the above use cases can be solved with it.
+
+.. note::
+
+   This is an experimental feature. Some details might change in later
+   versions of mypy. The final qualifiers are available in the
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Maybe instead of "the final qualifiers" write "the `final` decorator and `Final` pseudo-type"?

File: docs/source/final_attrs.rst [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+   ``typing_extensions`` package available on PyPI.
+
+Syntax variants
+***************
+
+The ``typing_extensions.Final`` qualifier indicates that a given name or
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: So I'm not sure that "qualifier" is the best term here, though I'm not sure that "pseudo-type" is any better. We probably should use the same word to describe "ClassVar" (see #5733).

File: docs/source/final_attrs.rst [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+attribute should never be re-assigned, re-defined, nor overridden. It can be
+used in one of these forms:
+
+
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Only one blank line.

File: docs/source/final_attrs.rst [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+* You can omit the type: ``ID: Final = 1``. Note that unlike for generic
+  classes this is *not* the same as ``Final[Any]``. Here mypy will infer
+  type ``int``.
+
+* In stub files you can omit the right hand side and just write
+  ``ID: Final[float]``.
+
+* Finally, you can define ``self.id: Final = 1`` (also with a type argument),
+  but this is allowed *only* in ``__init__`` methods.
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Why the limitation to `__init__` methods? In general we allow inferring or declaring the type of instance variables in any method (though IIRC a declaration must be on the *first* use). Is this because the implementation cannot ensure that there aren't any assignments in other places if `Final` isn't set in `__init__`?

File: docs/source/final_attrs.rst [hunk 6]
```diff
@@ -0,0 +1,267 @@
+Final names, methods and classes
+================================
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: I would call out somewhat early in this section that there's both `Final` (a type qualifier or pseudo-type) and `final` (a decorator).

File: docs/source/final_attrs.rst [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+     ID: Final = 1
+     ID: Final = 2  # Error: "ID" already declared as final
+
+     class SomeCls:
+         id: Final = 1
+         def __init__(self, x: int) -> None:
+             self.id: Final = x  # Error: "id" already declared in class body
+
+  Note that mypy has a single namespace for a class. So there can't be two
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Drop "two".

File: docs/source/final_attrs.rst [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+     ID = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         ID = 1
+         ID: Final = 2  # Error!
+
+* A final attribute declared in class body without an initializer must
+  be initialized in the ``__init__`` method (you can skip the initializer
+  in stub files):
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Again, why limit this to `__init__`?

File: docs/source/final_attrs.rst [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+     class SomeCls:
+         x: Final[int]
+         y: Final[int]  # Error: final attribute without an initializer
+         def __init__(self) -> None:
+             self.x = 1  # Good
+
+* ``Final`` can be only used as an outermost type in assignments or variable
+  annotations. using it in any other position is an error. In particular,
+  ``Final`` can't be used in annotations for function arguments:
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: This is also very similar to `ClassVar`, right?

File: docs/source/final_attrs.rst [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+  .. code-block:: python
+
+     x: List[Final[int]] = []  # Error!
+     def fun(x: Final[List[int]]) ->  None:  # Error!
+         ...
+
+* ``Final`` and ``ClassVar`` should not be used together. Mypy will infer
+  the scope of a final declaration automatically depending on whether it was
+  initialized in the class body or in ``__init__``.
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Hm, interesting. I can understand you don't want `ClassVar[Final]` or  `Final[ClassVar]` (though the latter kind of sounds all right :-).

File: docs/source/final_attrs.rst [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+* There must be *exactly one* assignment to a final attribute:
+
+  .. code-block:: python
+
+     ID = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         ID = 1
+         ID: Final = 2  # Error!
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Can you use a different name instead of `ID` for the class variable? Having the name the same distract from the point this example is making regarding duplication. I don't think you need an example showing that a final module global variable and a final class variable don't clash -- it's standard Python semantics (and common knowledge) that those namespaces don't interact.

File: docs/source/final_attrs.rst [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+     # file mod.py
+     from typing_extensions import Final
+
+     ID: Final = 1
+
+     class SomeCls:
+         ID: Final = 1
+
+         def meth(self) -> None:
+             self.ID = 2  # Error: can't assign to final attribute
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Again I think it would be less confusing if the module variable and the class variable had different names.

File: docs/source/final_attrs.rst [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+         ID: Final = 1  # OK
+
+     class Other(Base):
+         ID: Final = 2  # OK
+
+     class Combo(One, Other):  # Error: cannot override final attribute.
+         pass
+
+* Declaring a name as final only guarantees that the name wll not be re-bound
+  to other value, it doesn't make the value immutable. One can use immutable ABCs
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: other -> another

File: docs/source/final_attrs.rst [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+  and containers to prevent mutating such values:
+
+  .. code-block:: python
+
+     x: Final = ['a', 'b']
+     x.append('c')  # OK
+
+     y: Final[Sequance[str]] = ['a', 'b']
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Typo (Sequ*e*nce)

File: docs/source/final_attrs.rst [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+Final methods
+*************
+
+Like with attributes, sometimes it is useful to protect a method from
+overriding. In such situations one can use the ``typing_extensions.final``
+decorator:
+
+.. code-block:: python
+
+   from typing_extensions import final
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Add a comment emphasizing this uses a lower-case 'f'.

File: docs/source/final_attrs.rst [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+
+   class Base:
+       @final
+       def common_name(self) -> None:
+           ...
+
+   # 1000 lines later
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Drop.

File: docs/source/final_attrs.rst [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+  .. code-block:: python
+
+     x: Final = ['a', 'b']
+     x.append('c')  # OK
+
+     y: Final[Sequance[str]] = ['a', 'b']
+     y.append('x')  # Error: Sequance is immutable
+     z: Final = ('a', 'b')  # Also an option
+
+Final methods
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Ironic that here "Final" is capitalized even though in the code it uses a lower-case 'f', whereas in previous section headings "final" was not capitalized but the code used `Final`. :-) Maybe add "(the `@final` decorator)" to the title here?

File: docs/source/final_attrs.rst [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+   class Base:
+       @overload
+       def meth(self) -> None: ...
+       @overload
+       def meth(self, arg: int) -> int: ...
+       @final
+       def meth(self, x=None):
+           ...
+
+Final classes
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: Add "(the `@final` class decorator)"?

File: docs/source/final_attrs.rst [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+
+Here are some situations where using a final class may be useful:
+
+* A class wasn't designed to be subclassed. Perhaps subclassing does not
+  work as expected, or it's error-prone.
+* You want to retain the freedom to arbitrarily change the class implementation
+  in the future, and these changes might break subclasses.
+* You believe that subclassing would make code harder to understand or maintain.
+  For example, you may want to prevent unnecessarily tight coupling between
+  base classes and subclasses.
```
Comment (gvanrossum) [inline docs/source/final_attrs.rst]: These bullets all make almost the same point. Maybe we don't need to editorialize and users can figure out for themselves why final classes might be useful (they might even know the concept from another language).

File: docs/source/final_attrs.rst (hunk 2 above)
Comment (ilevkivskyi) [reply docs/source/final_attrs.rst]: We discussed this with Jukka, and IIRC he suggested just "qualifiers".

File: docs/source/final_attrs.rst (hunk 5 above)
Comment (ilevkivskyi) [reply docs/source/final_attrs.rst]: Because the instance attribute should be set only once, when the instance is created. I think we should also allow this in `__new__`, but this is blocked by https://github.com/python/mypy/issues/1021. I will try to reword here to make this clear and add a follow-up issue for `__new__`.

File: docs/source/final_attrs.rst (hunk 8 above)
Comment (ilevkivskyi) [reply docs/source/final_attrs.rst]: > Again, why limit this to `__init__`?

Same as above.

File: docs/source/final_attrs.rst (hunk 9 above)
Comment (ilevkivskyi) [reply docs/source/final_attrs.rst]: Yes.

File: docs/source/final_attrs.rst (hunk 11 above)
Comment (ilevkivskyi) [reply docs/source/final_attrs.rst]: OK, I will update these.

File: docs/source/final_attrs.rst (hunk 18 above)
Comment (ilevkivskyi) [reply docs/source/final_attrs.rst]: I am not sure this is needed, we can add this later.

File: docs/source/final_attrs.rst (hunk 19 above)
Comment (ilevkivskyi) [reply docs/source/final_attrs.rst]: IIRC, Jukka proposed to add some examples, and I think it probably makes sense to keep at least some of them.
